### PR TITLE
Check binary files that might be wrongly commited

### PR DESCRIPTION
**Pull request by privat**

### CODE REVIEW

File: misc/jenkins/checkbinaryfiles.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+err=0
+
+cd `git rev-parse --show-toplevel`
+
+err=0
+
+echo "checklicense $from (`git rev-parse "$from"`) .. $to (`git rev-parse "$to"`)"
+git diff --name-only $from..$to -z > checkbinaryfiles_files.out
+if test \! -s checkbinaryfiles_files.out; then
+	echo "No files files"
```
Comment (jcbrinfo) [inline misc/jenkins/checkbinaryfiles.sh]: ?

File: misc/jenkins/checkbinaryfiles.sh [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	exit 0
+fi
+xargs -0 file --mime < checkbinaryfiles_files.out | grep application > checkbinaryfiles_application.out
```
Comment (jcbrinfo) [inline misc/jenkins/checkbinaryfiles.sh]: Why `application`

File: misc/jenkins/checkbinaryfiles.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+err=0
+
+echo "checklicense $from (`git rev-parse "$from"`) .. $to (`git rev-parse "$to"`)"
+git diff --name-only $from..$to -z > checkbinaryfiles_files.out
+if test \! -s checkbinaryfiles_files.out; then
+	echo "No committed files"
+	exit 0
+fi
+xargs -0 file --mime < checkbinaryfiles_files.out | grep application > checkbinaryfiles_application.out
```
Comment (jcbrinfo) [inline misc/jenkins/checkbinaryfiles.sh]: Why `application`?

File: misc/jenkins/checkbinaryfiles.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+if test -s checkbinaryfiles_application.out; then
+	echo "These files type need a manual check:"
+	echo ""
+	cat checkbinaryfiles_application.out
+	echo ""
+	echo "Please double check that they are not wrongly committed files or that strange control character are not included in them."
+	exit 1
+else
+	echo "All committed files seems good"
```
Comment (jcbrinfo) [inline misc/jenkins/checkbinaryfiles.sh]: `All committed files seem good.`

File: misc/jenkins/checkbinaryfiles.sh (hunk 2 above)
Comment (privat) [inline misc/jenkins/checkbinaryfiles.sh]: a rule of thumb that encloses unwanted things like: application/x-executable application/x-sharedlib application/octet-stream etc.

**Comment by jcbrinfo** [comment]
You should `grep 'application/[^/]*$'` to avoid false positives (`application-server.nit: text/plain`…). But, even with this, you may have problems with the fact that the `application` namespace is not only for binary files: it is an “everything else” category. It includes things like programming languages (a subject of debate), mail formats, etc. Furthermore, your script does not look for binary image files, video files, audio files, etc.

**Comment by jcbrinfo** [comment]
Heuristic of Git:
> `builtin_diff()` calls `diff_filespec_is_binary()` which calls `buffer_is_binary()` which checks for any occurrence of a zero byte (NUL “character”) in the first 8000 bytes (or the entire length if shorter).
> — http://stackoverflow.com/a/6134127

**Comment by jcbrinfo** [comment]
```
$ grep -Paq '\x00' -- bin/nitc; printf '%d\n' "$?"
0
$ grep -Paq '\x00' -- src/nitc.nit; printf '%d\n' "$?"
1
```
Works with GNU Grep. I don’t know how to do the same with POSIX Grep, though.

**Comment by xymus** [comment]
+1

File: misc/jenkins/checkbinaryfiles.sh [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+from=${1:-origin/master}
+to=${2:-HEAD}
+
+err=0
+
+cd `git rev-parse --show-toplevel`
+
+err=0
+
+echo "checklicense $from (`git rev-parse "$from"`) .. $to (`git rev-parse "$to"`)"
```
Comment (xymus) [inline misc/jenkins/checkbinaryfiles.sh]: `checklicense`?

**Comment by privat** [comment]
test this please

**Pull request merged**